File: relax-core-spring-boot-starter/src/main/java/com/infilos/spring/utils/Respond.java
package com.infilos.spring.utils;

import com.fasterxml.jackson.annotation.*;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.vavr.control.Either;
import io.vavr.control.Try;
import org.springframework.http.*;

import java.io.ByteArrayOutputStream;
import java.util.Objects;

public final class Respond<T> {
  private static final Integer SuccedCode = 0;
  private static final Integer FailedCode = -1;
  private static final String BlankMessage = "";
  private final Integer code;
  private final T data;
  private final String message;

  @JsonCreator
  public Respond(
      @JsonProperty("code") int code,
      @JsonProperty("data") T data,
      @JsonProperty("message") String message) {
    this.code = code;
    this.data = data;
    this.message = message;
  }

  public static <T> Respond<T> succed() {
    return new Respond<>(SuccedCode, null, BlankMessage);
  }

  public static <T> Respond<T> succed(T data) {
    return new Respond<>(SuccedCode, data, BlankMessage);
  }

  public static <T> Respond<T> succed(int code, T data) {
    return new Respond<>(code, data, BlankMessage);
  }

  public static <T> Respond<T> succed(int code) {
    return new Respond<>(code, null, BlankMessage);
  }

  public static <T> Respond<T> failed(String message) {
    return new Respond<>(FailedCode, null, message);
  }

  public static <T> Respond<T> failed(int code, String message) {
    return new Respond<>(code, null, message);
  }

  public static <T> Respond<T> failed(Throwable cause) {
    return new Respond<>(
        FailedCode,
        null,
        String.format(
            "%s(%s)",
            cause.getClass().getSimpleName(),
            Objects.isNull(cause.getMessage()) ? "" : cause.getMessage()));
  }

  public static <T> Respond<T> failed(int code, Throwable cause) {
    return new Respond<>(
        code,
        null,
        String.format(
            "%s(%s)",
            cause.getClass().getSimpleName(),
            Objects.isNull(cause.getMessage()) ? "" : cause.getMessage()));
  }

  public static <T> Respond<T> failed() {
    return new Respond<>(FailedCode, null, "Unknown failure");
  }

  public static <T> Respond<T> of(Either<?, T> either) {
    if (either.isLeft()) {
      Object error = either.getLeft();
      if (error instanceof String) return failed((String) error);
      else if (error instanceof Throwable) return failed((Throwable) error);
      else return failed(error.toString());
    } else {
      return succed(either.get());
    }
  }

  public static ResponseEntity<byte[]> ofBytes(
      HttpStatus status, HttpHeaders headers, byte[] bytes) {
    return new ResponseEntity<>(bytes, headers, status);
  }

  public static ResponseEntity<byte[]> ofFileBytes(String name, ByteArrayOutputStream stream) {
    HttpHeaders headers = new HttpHeaders();
    headers.add(HttpHeaders.CONTENT_DISPOSITION, String.format("attachment; filename=%s", name));
    headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM_VALUE);

    return ofBytes(HttpStatus.OK, headers, stream.toByteArray());
  }

  public static ResponseEntity<byte[]> ofFileBytes(String name, String format, byte[] bytes) {
    String mediaType =
        Try.of(() -> MediaType.parseMediaType(format))
            .map(MediaType::toString)
            .toJavaOptional()
            .orElse(
                MediaTypeFactory.getMediaType(name)
                    .map(MediaType::toString)
                    .orElse(MediaType.APPLICATION_OCTET_STREAM_VALUE));

    HttpHeaders headers = new HttpHeaders();
    headers.add(HttpHeaders.CONTENT_DISPOSITION, String.format("attachment; filename=%s", name));
    headers.add(HttpHeaders.CONTENT_TYPE, mediaType);

    return ofBytes(HttpStatus.OK, headers, bytes);
  }

  @JsonProperty
  public Integer getCode() {
    return code;
  }

  @JsonProperty
  public T getData() {
    return data;
  }

  @JsonProperty
  public String getMessage() {
    return message;
  }

  @Override
  public String toString() {
    return String.format("Respond(code=%d, message=%s, data=%s)", code, message, data);
  }

  @JsonIgnore
  public boolean isSucced() {
    return SuccedCode.equals(code);
  }

  @JsonIgnore
  public boolean isSuccedWithData() {
    return isSucced() && Objects.nonNull(data);
  }

  /** Convert this Respond to json ResponseEntity. */
  @JsonIgnore
  public ResponseEntity<byte[]> asBytesResponse(ObjectMapper objectMapper) {
    HttpHeaders headers = new HttpHeaders();
    headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);

    byte[] bytes;
    try {
      bytes = objectMapper.writeValueAsBytes(this);
    } catch (JsonProcessingException e) {
      bytes =
          String.format(
                  "{\"code\":-1,\"message\":\"Serialize %s failed: %s\"}",
                  this.toString(), e.getMessage())
              .getBytes();
    }

    return new ResponseEntity<>(bytes, headers, HttpStatus.OK);
  }
}

File: relax-scala-spring-boot-starter/src/main/scala/com/infilos/spring/utils/Respond.scala
package com.infilos.spring.utils

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.databind.ObjectMapper
import com.infilos.spring.utils.Respond.SuccedCode
import org.springframework.http._

import java.util.Optional
import scala.util.{Failure, Success, Try}

final case class Respond[+T](code: Int, data: T, message: String) {

  @JsonIgnore
  def isSuccedWithData: Boolean = isSucced && data != null

  @JsonIgnore
  def isSucced: Boolean = code == SuccedCode

  @JsonIgnore
  def asBytesResponse(objectMapper: ObjectMapper): ResponseEntity[Array[Byte]] = {
    val headers = new HttpHeaders()
    headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)

    var bytes: Array[Byte] = null
    try {
      bytes = objectMapper.writeValueAsBytes(this)
    } catch {
      case e: Throwable =>
        bytes = String.format(s"{\"code\":-1,\"$message\":\"Serialize $toString failed: ${e.getMessage}\"}", this.toString, e.getMessage).getBytes
    }

    new ResponseEntity[Array[Byte]](bytes, headers, HttpStatus.OK);
  }

  override def toString: String =
    s"Respond(code=$code, data=$data, message=$message)"

  override def equals(obj: Any): Boolean = obj match {
    case that: Respond[_] =>
      this.canEqual(that) &&
        this.code == that.code &&
        this.data == that.data &&
        this.message == that.message
    case _ => false
  }

  override def canEqual(that: Any): Boolean = that.isInstanceOf[Respond[_]]

  override def hashCode(): Int = (code, data, message).##
}

object Respond {

  val SuccedCode: Int = 0
  val FailedCode: Int = -1
  val BlankMessage: String = ""

  def succed[T](): Respond[T] =
    Respond(SuccedCode, null.asInstanceOf[T], BlankMessage)

  def succed[T](code: Int, data: T): Respond[T] =
    Respond(code, data, BlankMessage)

  def succed[T](code: Int): Respond[T] =
    Respond(code, null.asInstanceOf[T], BlankMessage)

  def failed[T](code: Int, error: String): Respond[T] =
    Respond(code, null.asInstanceOf[T], error)

  def failed[T](code: Int, cause: Throwable, retry: Boolean = false): Respond[T] =
    Respond(
      code,
      null.asInstanceOf[T],
      s"${cause.getClass.getSimpleName}(${if (cause.getMessage == null) "" else cause.getMessage})"
    )

  def failed[T](): Respond[T] =
    Respond(FailedCode, null.asInstanceOf[T], BlankMessage)

  def of[T](either: Either[_, T]): Respond[T] = either match {
    case Left(value) => value match {
      case error: String => failed(error)
      case error: Throwable => failed(error)
      case error => failed(error.toString)
    }
    case Right(value) => succed(value)
  }

  def succed[T](data: T): Respond[T] =
    Respond(SuccedCode, data, BlankMessage)

  def failed[T](error: String): Respond[T] =
    Respond(FailedCode, null.asInstanceOf[T], error)

  def failed[T](cause: Throwable): Respond[T] = {
    Respond(
      FailedCode,
      null.asInstanceOf[T],
      s"${cause.getClass.getSimpleName}(${if (cause.getMessage == null) "" else cause.getMessage})"
    )
  }

  def ofFileBytes(filename: String, bytes: Array[Byte]): ResponseEntity[Array[Byte]] = {
    val headers = new HttpHeaders()
    headers.add(HttpHeaders.CONTENT_DISPOSITION, s"attachment; filename=$filename")
    headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM_VALUE)

    ofBytes(HttpStatus.OK, headers, bytes)
  }

  def ofFileBytes(filename: String, format: String, bytes: Array[Byte]): ResponseEntity[Array[Byte]] = {
    val mediaType: Optional[String] = Try(MediaType.parseMediaType(format)) match {
      case Success(value) => Optional.ofNullable(value).map(_.toString)
      case Failure(_) => MediaTypeFactory.getMediaType(filename).map(_.toString)
    }

    val headers = new HttpHeaders()
    headers.add(HttpHeaders.CONTENT_DISPOSITION, s"attachment; filename=$filename")
    headers.add(HttpHeaders.CONTENT_TYPE, mediaType.orElse(MediaType.APPLICATION_OCTET_STREAM_VALUE))

    ofBytes(HttpStatus.OK, headers, bytes)
  }

  def ofBytes(status: HttpStatus, headers: HttpHeaders, bytes: Array[Byte]): ResponseEntity[Array[Byte]] = {
    new ResponseEntity[Array[Byte]](bytes, headers, status)
  }
}
